lib/rpc/ProviderHealthStateDynamoDbRepository.ts

```diff
@@ -0,0 +1,86 @@
+import { ProviderHealthStateRepository } from './ProviderHealthStateRepository'
+import Logger from 'bunyan'
+import { DynamoDB } from 'aws-sdk'
+import { DocumentClient } from 'aws-sdk/clients/dynamodb'
+import { ProviderHealthState } from './ProviderHealthState'
+
+// Table item assignment
+const UPDATE_EXPRESSION = `SET 
+  #healthiness = :healthiness,
+  #ongoingAlarms = :ongoingAlarms,
+  #version = :version`
+
+// Table column names
+const EXPRESSION_ATTRIBUTE_NAMES = {
+  '#healthiness': 'healthiness',
+  '#ongoingAlarms': 'ongoingAlarms',
+  '#version': 'version',
+}
+
+const CONDITION_EXPRESSION = '#version = :baseVersion'
+
+export class ProviderHealthStateDynamoDbRepository implements ProviderHealthStateRepository {
+  private ddbClient: DynamoDB.DocumentClient
+
+  constructor(private dbTableName: string, private log: Logger) {
+    this.ddbClient = new DynamoDB.DocumentClient()
+  }
+
+  async read(providerId: string): Promise<ProviderHealthState | null> {
+    const getParams = {
+      TableName: this.dbTableName,
+      Key: { chainIdProviderName: providerId },
+    }
+    try {
+      const result = await this.ddbClient.get(getParams).promise()
+      const item = result.Item
+      if (item === undefined) {
+        this.log.debug(`No health state found for ${providerId}`)
+        return null
+      }
+      return {
+        healthiness: item.healthiness,
+        ongoingAlarms: item.ongoingAlarms,
+        version: item.version,
+      }
+    } catch (error: any) {
+      this.log.error(`Failed to read health state from DB: ${JSON.stringify(error)}`)
+      throw error
+    }
+  }
+
+  async write(providerId: string, state: ProviderHealthState): Promise<void> {
+    const putParams: DocumentClient.PutItemInput = {
+      TableName: this.dbTableName,
+      Item: {
+        chainIdProviderName: providerId,
+        healthiness: state.healthiness,
+        ongoingAlarms: state.ongoingAlarms,
+        version: state.version,
+      },
+    }
+    await this.ddbClient.put(putParams).promise()
+    return
+  }
+
+  async update(providerId: string, state: ProviderHealthState): Promise<void> {
+    const updateParams: DocumentClient.UpdateItemInput = {
+      TableName: this.dbTableName,
+      Key: { chainIdProviderName: providerId },
+      UpdateExpression: UPDATE_EXPRESSION,
+      ExpressionAttributeNames: EXPRESSION_ATTRIBUTE_NAMES,
+      ExpressionAttributeValues: this.getExpressionAttributeValues(state),
+      ConditionExpression: CONDITION_EXPRESSION,
+    }
+    await this.ddbClient.update(updateParams).promise()
+  }
+
+  private getExpressionAttributeValues(state: ProviderHealthState): { [key: string]: any } {
+    let attributes: { [key: string]: any } = {}
+    attributes[':healthiness'] = state.healthiness
+    attributes[':ongoingAlarms'] = state.ongoingAlarms
+    attributes[':baseVersion'] = state.version - 1
```

Comment:
why do we need to the base version and nee version? also if the differene is only one, do we ever need it stored?

Reply:
I should've added more comments in the code. It's used for dealing with write conflicts, that is, two fallback lambdas trying to write to the same DB primary key at the same time.

All entry updates are designed to increase the version number by 1, so we expect the existing DB version number to be minus 1 of the version number in the update. However, if it's not the case, it means a write conflict has occurred and the latter update should cancel and retry with another lambda call.

I've verified that, if the lambda throws exception, AWS will retry another 2 times with the same input (alarm event object), with each having a delay of about 60 seconds. So in the retry, it should be able to finish the write.

Reply:
If the conditional expression `'#version = :baseVersion'` evaluates to false, `this.ddbClient.update()` will throw an exception

```
ConditionalCheckFailedException: The conditional request failed
```
There's no `try...catch` to catch this error so it will pop up as a lambda invocation error.

Optimistic locking is a better choice here because the probability of having concurrent update is rare in our use case. The total amount of alarm state changes regarding a chain-provider pair is small and will probably not even causing the spawn of another lambda (however, if it happens and we do nothing about, some provider might never recover). In comparison, adding a lock (pessimistic locking) is a much less efficient way. It's only be worth it if we want absolute transactional updates or the frequency of conflict is high...both are not the case here.

Reply:
i still dont think its a good solution, but dont want to block you

Reply:
Can you share why you don't think it's a good solution?